### DISCUSSION TITLE
ESP32 & CH340 variants

### DISCUSSION THREAD

**Discussion by pheetr** [awawa-dev/HyperHDR#159] [Q&A]
Hi,
I'm trying to get a SK6812 led strip working with a RPi Zero 2 and I've run into some challenges. My initial attempt was without a level shifter or an ESP32/ESP8266 board so now I'm in the process of acquiring these. While I've found the level shifter suggested in the wiki, I need some help with picking the right ESP32.
Does it matter which CH340 variant the board comes with? While the wiki specifically recommends 340G, most options I'm finding at the moment seem to have 340C implemented and one I'm leaning toward has the 340K. 
The [product page for CH340](http://wch-ic.com/products/CH340.html) mentions one difference, that while the C and K models include a clock G doesn't and needs an external one. I'm new to these components, so I'm not sure whether this makes a difference for our use case. 

I'm looking at the following boards specifically. Would these work? If yes, is one better than the other, or if not, would someone be able to recommend a suitable one? (ideally shipping within EU - don't want to wait too long :)

[DFRobot FireBeetle ESP32-E](https://thepihut.com/products/firebeetle-esp32-e-iot-microcontroller-supports-wi-fi-bluetooth?variant=39493908201667) with CH340K according to [schematic](https://dfimg.dfrobot.com/nobody/wiki/fd28d987619c16281bdc4f40990e5a1c.PDF).
[AZDelivery ESP32 Lolin LOLIN32](https://www.ebay.ie/itm/144123756354) with CH340C according to image.

As an aside, I do have a Wemos D1 mini at home that is meant for another project. I could temporarily repurpose it, but it also has a CH340C, not the G model.

Thanks.

**Comment by awawa-dev** [comment]
Hi
There are some reports about problems with CH340 variants recently. Not sure which particular model is causing the problems because boards were advertised as including CH340G but the chip looked differently and its markings were removed. Seems it has some hardware limitation compared to CH340G and it is detected by the Linux driver.. I would stick with ch340G if you are sure about supplier or even would consider CH9201x. You can find one source from I order some mini boards (ESP32, Ali so the shipping could take long) but as you know the offer could changed: [LINK](https://github.com/awawa-dev/HyperHDR/issues/129) (CH9201x already replaced cp2104 visible on the pictures, cp2104 is also good)

**Comment by awawa-dev** [comment]
If you have CH340C maybe connect it to the linux and see the dmesg logs to check if it has hardware break control or only software simulation.

**Reply by pheetr** [reply]
Thanks for the quick replies.
I have looked for CH9201x and cp2104 as well, but wasn't having much luck locating a board that would ship more locally. I've found a bunch with cp2104 in the title but with cp2102 in the pictures and didn't trust them based on that. There seemed to be more CH340 around, hence why I started mainly looking into them. I'll take another look at the other options though, seeing that the situation around CH340 isn't too clear apparently.  

In the meantime I've connected the Wemos D1 mini with the CH340C to a linux box and run dmesg.
I have the following output but I'm not sure what to look for in terms of hardware break control. Do I need to run a more specific command perhaps?
```
[1997121.730398] usb 3-1: new full-speed USB device number 2 using xhci_hcd
[1997121.879314] usb 3-1: New USB device found, idVendor=1a86, idProduct=7523, bcdDevice= 2.64
[1997121.879315] usb 3-1: New USB device strings: Mfr=0, Product=2, SerialNumber=0
[1997121.879316] usb 3-1: Product: USB Serial
[1997121.901129] usbcore: registered new interface driver usbserial_generic
[1997121.901135] usbserial: USB Serial support registered for generic
[1997121.902065] usbcore: registered new interface driver ch341
[1997121.902075] usbserial: USB Serial support registered for ch341-uart
[1997121.903526] usb 3-1: ch341-uart converter now attached to ttyUSB0
```

**Reply by awawa-dev** [reply]
That fatal message should appear here if that was the case described by other users. So it's rather OK. Using external (340G) or internal (340C) clock should not make any difference as it should be 12MHz in both cases. in fact external version is more dependent on the overall board quality.

**Reply by awawa-dev** [reply]
I found something in the documentation about version C and K: slightly increased error rate. Not sure how it will behave at 2Mb. Fortunately we have data integration check. Reception error rate is the same.

```CH340  has  integrated  separate  transmit-receive  buffer  and  supports  simplex,  half-duplex  and  full  duplex 
UART  communication.  Serial  data  contains  one  low-level  start  bit  ,  5,  6,  7  or  8  data  bits  and  1  or  2 
high-level stop bits, supports  odd/even/mark/space check. CH340 supports common baud rate: 50, 75, 100, 
110, 134.5, 150, 300, 600, 900, 1200, 1800, 2400, 3600, 4800, 9600, 14400, 19200, 28800, 33600, 38400, 
56000, 57600, 76800, 115200, 128000, 153600, 230400, 460800, 921600, 1500000, 2000000 etc.    
The  baud  rate  error  of  CH340  UART  reception  is  about  2%,  the  baud  rate  error  of  CH340G/CH340T/CH340R 
UART transmission is less than 0.3%, less than 1% for CH340C/CH340N/CH340K/CH340E/CH340B```

**Reply by pheetr** [reply]
Interesting. Thanks for looking into this further.

In the meantime I've ordered both the board you've linked above that should have the CH9201x as well as the DFRobot board with the 340K, which should get to me much quicker. :)

**Reply by pheetr** [reply]
In the meantime, the board I've ordered from the same link that you had used has arrived and it has CH9102x. Is this the chip you referenced as well? You wrote CH9201x, but a search for that didn't give me results other than on hyperhdr page, so I assume it was a typo and the chip is in fact CH9102x?
At any rate, I'll attempt the build using HyperSPI. That should be the best way for PiZero2 + SK6812, right?

**Reply by awawa-dev** [reply]
Yes, it's my typo: the serial chip is CH9102x of course. If you can put your ESP32 close to the Rpi then definitely go with HyperSPI (>25Mb with minimal protocol overhead resulting in bit higher LED refresh-rate), otherwise recommend HyperSerialEsp32 (2Mb, could be modify to 4Mb for CH9102x , rather high USB related overhead compared to pure SPI).

**Reply by pheetr** [reply]
Thanks for confirming. Yes, the ESP32 will be right next to the Pi, so I'll go for HyperSPI. 

Apologies for the noob question in advance, but this is my first project with ESP32 or the Arduino IDE. I've followed the steps on the HyperSPI page and the linked espressif install guide. In Arduino IDE:
- Installed the two libraries
- set the board to 'MH ET Live ESP32Minikit'
- loaded 'HyperSPI_esp32_SK6812NeutralWhite.ino' in IDE
- selected COM4
- executed Upload

It finished with 'Done uploading', and I see no obvious errors in [the output](https://gist.github.com/pheetr/24bf2d51082cc985c1ba2885b193b830), so I assume it went fine.
What is the best way to check if it is in fact working as intended?

**Reply by awawa-dev** [reply]
Very good question: it is better to be sure that it is loaded and working properly than to search unnecessarily for possible errors in the installation later. You can use any serial port client, even Arduino IDE serial console is enough: just set communication speed at 115200 and you should get some readings (restart with a ESP reset button should trigger even more info on the serial port output).

**Reply by awawa-dev** [reply]
BTW you can use esphome-flasher & pre-build images available in the HyperSPI release section: using Arduino for building HyperSPI from sources is not necessary.

**Reply by pheetr** [reply]
Thanks, using the Arduino console I got the following after pressing the reset button:
```
r.⸮⸮⸮⸮ (POWERON_RESET),boot:0⸮3 (SPI_FAST_FLASH_BOOT)
config⸮⸮ 0, SPIWP:0xee
clk_drv:0x00,\E⸮⸮⸮0x00,d_drv:0x00,cs0_drv:0x⸮bB⸮}⸮⸮v:0x00,wp_drv:0x00
mode:DIO, ,⸮k⸮⸮div:1
load:0x3fff0030,len:1184
load:0x40078000,⸮⸮⸮12804
ho 0 tail 12 room 4
load:0x40080⸮bb⸮⸮⸮3032
entry 0x400805e4

Welcome!
Awa SPI driver.
Color mode: RGBW (neutral white)
First LED: enabled
HyperSPI version 1.
Statistics for the last full 1 second cycle.
Frames per second: 0
Good frames: 0
Bad frames:  0
```
Based on the above, I assume things are ok so far.

**Reply by d22b** [reply]
Hi. I build up up HDR from scratch with v19 and tested the build (direct connect to raspi) to make sure everything works as designed. So now I would like to use the option with hyperserial and a dedicated led controller board with an equipped S2 Mini ontop.

I briefly followed your instructions and build the firmware from the GitHub repository. So far, so good. But now I am stucked: I tried to flash the image to the S2 mini with the esp flasher. I am not able to flash the image. What I am tried:
- boot mode (press 0 and the rst) 
- different flashers 

is there a magic trick to bring the image to the board?

**Reply by awawa-dev** [reply]
I think creating new topic in HyperSerialESP32 would be easier to find for me and others. I've updated recently manual for ESP32-S2 and added more detailed manual for this board. Check it out: https://github.com/awawa-dev/HyperSerialESP32#flashing and report if it works for you.
